In [1]:
from transformers import pipeline

https://huggingface.co/distilbert/distilbert-base-uncased

In [2]:
mask_pipeline = pipeline("fill-mask", model="distilbert-base-uncased")

mask_pipeline

/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


config.json:   0%|          | 0.00/483 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/268M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

Device set to use cpu


In [3]:
text = "He loves to play [MASK] on weekends."

text

'He loves to play [MASK] on weekends.'

In [4]:
mask_pipeline(text)

[{'score': 0.332697331905365,
  'token': 5439,
  'token_str': 'golf',
  'sequence': 'he loves to play golf on weekends.'},
 {'score': 0.0930471271276474,
  'token': 11662,
  'token_str': 'poker',
  'sequence': 'he loves to play poker on weekends.'},
 {'score': 0.0785880759358406,
  'token': 5093,
  'token_str': 'tennis',
  'sequence': 'he loves to play tennis on weekends.'},
 {'score': 0.0672822892665863,
  'token': 3598,
  'token_str': 'baseball',
  'sequence': 'he loves to play baseball on weekends.'},
 {'score': 0.05059058591723442,
  'token': 7433,
  'token_str': 'chess',
  'sequence': 'he loves to play chess on weekends.'}]

In [5]:
text = "She found the book very [MASK]."

mask_pipeline(text)

[{'score': 0.0948922410607338,
  'token': 5875,
  'token_str': 'interesting',
  'sequence': 'she found the book very interesting.'},
 {'score': 0.05977049842476845,
  'token': 17160,
  'token_str': 'fascinating',
  'sequence': 'she found the book very fascinating.'},
 {'score': 0.05858134850859642,
  'token': 14036,
  'token_str': 'entertaining',
  'sequence': 'she found the book very entertaining.'},
 {'score': 0.04417252913117409,
  'token': 19142,
  'token_str': 'amusing',
  'sequence': 'she found the book very amusing.'},
 {'score': 0.043478529900312424,
  'token': 16004,
  'token_str': 'appealing',
  'sequence': 'she found the book very appealing.'}]

In [6]:
text = "The weather today is [MASK]."

mask_pipeline(text, top_k=3)

[{'score': 0.12658391892910004,
  'token': 10256,
  'token_str': 'mild',
  'sequence': 'the weather today is mild.'},
 {'score': 0.09278527647256851,
  'token': 24706,
  'token_str': 'cloudy',
  'sequence': 'the weather today is cloudy.'},
 {'score': 0.04285835102200508,
  'token': 24166,
  'token_str': 'stormy',
  'sequence': 'the weather today is stormy.'}]

In BERT-based models, the [CLS] token is always placed at the start of every input sentence.

After processing, the output embedding of the [CLS] token is used as a summary representation of the whole sequence.

In [7]:
text =  "She wore a [MASK] dress and a [MASK] hat."

mask_pipeline(text)

[[{'score': 0.07334253191947937,
   'token': 2317,
   'token_str': 'white',
   'sequence': '[CLS] she wore a white dress and a [MASK] hat. [SEP]'},
  {'score': 0.056657370179891586,
   'token': 2417,
   'token_str': 'red',
   'sequence': '[CLS] she wore a red dress and a [MASK] hat. [SEP]'},
  {'score': 0.046419136226177216,
   'token': 2304,
   'token_str': 'black',
   'sequence': '[CLS] she wore a black dress and a [MASK] hat. [SEP]'},
  {'score': 0.0432610809803009,
   'token': 2630,
   'token_str': 'blue',
   'sequence': '[CLS] she wore a blue dress and a [MASK] hat. [SEP]'},
  {'score': 0.03892374038696289,
   'token': 5030,
   'token_str': 'wedding',
   'sequence': '[CLS] she wore a wedding dress and a [MASK] hat. [SEP]'}],
 [{'score': 0.18764649331569672,
   'token': 13137,
   'token_str': 'straw',
   'sequence': '[CLS] she wore a [MASK] dress and a straw hat. [SEP]'},
  {'score': 0.15933720767498016,
   'token': 11762,
   'token_str': 'cowboy',
   'sequence': '[CLS] she wore a 

In [8]:
text = "On vacation, I like to explore the [MASK] and try a new [MASK]."

mask_pipeline(text)

[[{'score': 0.05310136079788208,
   'token': 10833,
   'token_str': 'countryside',
   'sequence': '[CLS] on vacation, i like to explore the countryside and try a new [MASK]. [SEP]'},
  {'score': 0.0361083522439003,
   'token': 4020,
   'token_str': 'mountains',
   'sequence': '[CLS] on vacation, i like to explore the mountains and try a new [MASK]. [SEP]'},
  {'score': 0.03372699022293091,
   'token': 4153,
   'token_str': 'ocean',
   'sequence': '[CLS] on vacation, i like to explore the ocean and try a new [MASK]. [SEP]'},
  {'score': 0.03117975778877735,
   'token': 19350,
   'token_str': 'outdoors',
   'sequence': '[CLS] on vacation, i like to explore the outdoors and try a new [MASK]. [SEP]'},
  {'score': 0.029092565178871155,
   'token': 3509,
   'token_str': 'beach',
   'sequence': '[CLS] on vacation, i like to explore the beach and try a new [MASK]. [SEP]'}],
 [{'score': 0.05544567108154297,
   'token': 6172,
   'token_str': 'adventure',
   'sequence': '[CLS] on vacation, i like

Even if the training data used for this model could be characterized as fairly neutral, this model can have biased predictions

In [9]:
text = "The gentleman worked as a [MASK]."

mask_pipeline(text)

[{'score': 0.12069959193468094,
  'token': 10533,
  'token_str': 'carpenter',
  'sequence': 'the gentleman worked as a carpenter.'},
 {'score': 0.09845943748950958,
  'token': 20987,
  'token_str': 'blacksmith',
  'sequence': 'the gentleman worked as a blacksmith.'},
 {'score': 0.06329609453678131,
  'token': 22701,
  'token_str': 'tailor',
  'sequence': 'the gentleman worked as a tailor.'},
 {'score': 0.055799614638090134,
  'token': 7500,
  'token_str': 'farmer',
  'sequence': 'the gentleman worked as a farmer.'},
 {'score': 0.04521230235695839,
  'token': 7805,
  'token_str': 'clerk',
  'sequence': 'the gentleman worked as a clerk.'}]

In [10]:
text = "The lady worked as a [MASK]."

mask_pipeline(text)

[{'score': 0.17885392904281616,
  'token': 10850,
  'token_str': 'maid',
  'sequence': 'the lady worked as a maid.'},
 {'score': 0.08888030797243118,
  'token': 6821,
  'token_str': 'nurse',
  'sequence': 'the lady worked as a nurse.'},
 {'score': 0.07091611623764038,
  'token': 22583,
  'token_str': 'housekeeper',
  'sequence': 'the lady worked as a housekeeper.'},
 {'score': 0.04480187967419624,
  'token': 5660,
  'token_str': 'cook',
  'sequence': 'the lady worked as a cook.'},
 {'score': 0.039416369050741196,
  'token': 10533,
  'token_str': 'carpenter',
  'sequence': 'the lady worked as a carpenter.'}]

* https://huggingface.co/docs/transformers/model_doc/auto
* https://huggingface.co/bert-base-uncased

In [11]:
from transformers import AutoModelForMaskedLM, AutoTokenizer

In [12]:
BERT_MODEL = "bert-base-uncased"

BERT_MODEL

'bert-base-uncased'

In [13]:
tokenizer = AutoTokenizer.from_pretrained(BERT_MODEL)

tokenizer

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/570 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

BertTokenizerFast(name_or_path='bert-base-uncased', vocab_size=30522, model_max_length=512, is_fast=True, padding_side='right', truncation_side='right', special_tokens={'unk_token': '[UNK]', 'sep_token': '[SEP]', 'pad_token': '[PAD]', 'cls_token': '[CLS]', 'mask_token': '[MASK]'}, clean_up_tokenization_spaces=False, added_tokens_decoder={
	0: AddedToken("[PAD]", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
	100: AddedToken("[UNK]", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
	101: AddedToken("[CLS]", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
	102: AddedToken("[SEP]", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
	103: AddedToken("[MASK]", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
}
)

In [14]:
model = AutoModelForMaskedLM.from_pretrained(BERT_MODEL)

model

model.safetensors:   0%|          | 0.00/440M [00:00<?, ?B/s]

Some weights of the model checkpoint at bert-base-uncased were not used when initializing BertForMaskedLM: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight', 'cls.seq_relationship.bias', 'cls.seq_relationship.weight']
- This IS expected if you are initializing BertForMaskedLM from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForMaskedLM from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).


BertForMaskedLM(
  (bert): BertModel(
    (embeddings): BertEmbeddings(
      (word_embeddings): Embedding(30522, 768, padding_idx=0)
      (position_embeddings): Embedding(512, 768)
      (token_type_embeddings): Embedding(2, 768)
      (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
      (dropout): Dropout(p=0.1, inplace=False)
    )
    (encoder): BertEncoder(
      (layer): ModuleList(
        (0-11): 12 x BertLayer(
          (attention): BertAttention(
            (self): BertSdpaSelfAttention(
              (query): Linear(in_features=768, out_features=768, bias=True)
              (key): Linear(in_features=768, out_features=768, bias=True)
              (value): Linear(in_features=768, out_features=768, bias=True)
              (dropout): Dropout(p=0.1, inplace=False)
            )
            (output): BertSelfOutput(
              (dense): Linear(in_features=768, out_features=768, bias=True)
              (LayerNorm): LayerNorm((768,), eps=1e-12, elementwi


- Run this with the following sentences

` She wore an elegant [MASK] to the event.`

`He wore a elegant [MASK] to the event.`

In [15]:
text = "He wore an elegant [MASK] to the event."

text

'He wore an elegant [MASK] to the event.'

In [16]:
input_ids = tokenizer.encode(text, return_tensors="pt")

input_ids

tensor([[  101,  2002,  5078,  2019, 11552,   103,  2000,  1996,  2724,  1012,
           102]])

##### finds the indices of masked tokens in the input_ids (there is just one token_id = 103)

In [17]:
mask_token_indices = (input_ids == tokenizer.mask_token_id).nonzero(as_tuple=True)

mask_token_indices

(tensor([0]), tensor([5]))

#####  iterates over the indices of masked tokens in the input obtains the model predictions for each masked token and appends the decoded word corresponding to the highest probability token to predicted_words

In [18]:
import torch

predicted_words = []

# Iterate over all the masks present in the input
for i in range(mask_token_indices[0].shape[0]):
    mask_token_index = mask_token_indices[1][i]

    predictions = model(input_ids)

    # Get logits for the masked word prediction
    # Logits shape [batch_size, length_of_input, num_predicted_words]
    predicted_token_logits = predictions.logits[0, mask_token_index, :]

    # Apply softmax to get probabilities
    predicted_token_probabilities = torch.softmax(predicted_token_logits, dim=-1)

    # Find the predicted token id with the highest probability
    predicted_token_ids = torch.argmax(predicted_token_probabilities, dim=-1)

    # Decode the predicted token id to a word
    predicted_word = tokenizer.decode([predicted_token_ids.item()])

    # Check if the prediction is not null or empty
    if predicted_word and predicted_word.strip():
        predicted_words.append(predicted_word)
    else:
        predicted_words.append("[MASK]")  # or some other placeholder for null prediction

In [19]:
predicted_words

['suit']

In [20]:
for predicted_word in predicted_words:
    text = text.replace("[MASK]", predicted_word, 1)

print(text)

He wore an elegant suit to the event.
